src/gpu_sysman_test.c

```diff
@@ -64,9 +69,15 @@
  * - Plugin init, shutdown and re-init works without problems
  */
 
-#define KERNEL_LINUX 1
-#define FP_LAYOUT_NEED_NOTHING 1
-#include "gpu_sysman.c" /* test this */
+/* utilities needed from collectd core */
+#include "daemon/metric.c"
+#include "daemon/utils_cache_mock.c"
+#include "utils/common/common.c"
+#include "utils/metadata/meta_data.c"
+#include "utils/strbuf/strbuf.c"
+#include "utils/utf8/utf8.c"
```

Comment:
Why doesn't `gpu_sysman.c` include these?

Reply:
Because `gpu_sysman.c` is a collect plugin [1], whereas  `gpu_sysman_test.c` is a separate binary.

[1] i.e. collectd provides the symbols:
```
$ strip -s collectd
$ readelf -sW collectd | grep -v UND | awk '/FUNC *GLOBAL/{print $8}' | cut -b-7 | sort | uniq -c | sort -nr
     58 plugin_
     19 meta_da
     14 metric_
     13 uc_meta
     12 uc_get_
      9 strbuf_
...
```

The reasons why test binary includes all the necessary code directly, instead of linking to archives built during collectd build, are several:
* With all used code being visible to GCC, it can do proper whole-program analysis
* I'm testing the code in question with more strict options i.e. I will improve quality also in them
* I run these tests separately from the collectd build

Reply:
Oh! That is unexpected. I'd actually much prefer to consistently build tests by linking to modules that are not under test.

* Whole program analysis: Why is this needed? My understanding is that whole program analysis is primarily done to optimize the code, and I don't particularly care about the performance of unit tests.
* Stricter cflags: We can add an additional build to the CI system that builds everything with stricter flags. We can probably also run tests with ASAN instead of Valgrind.
* Alternative build method: I'm not keen on maintaining alternative, bespoke way of building parts of the tree. In other words, maintainability is more important to me than supporting such one-off use cases.

Reply:
> * Whole program analysis: Why is this needed? My understanding is that whole program analysis is primarily done to optimize the code, and I don't particularly care about the performance of unit tests.

Warnings analysis is part of compiler optimization passes.  If you do not enable optimizations, code checks can miss a lot of things.  And with contents of all functions visible to compiler, it can do better checks (along with those optimizations), for things actually used in the test.


> * Stricter cflags: We can add an additional build to the CI system that builds everything with stricter flags.

Enabling all the flags I'm interested about for whole collectd, would require a lot of fixes to rest of the code first.  Now I've needed to fix only things that are included directly to the tester.

(I can work on fixing those for collectd core, but I do not feel comfortable doing it for plugins that I'm not using, unless my changes are really trivial/obvious improvements.)

> We can probably also run tests with ASAN instead of Valgrind.

Yes, that would be really nice!


> * Alternative build method: I'm not keen on maintaining alternative, bespoke way of building parts of the tree. 

Sorry, I forgot to mention most important reason of including just the needed C-files instead of using the archives.

To do the checks requires by tests, I need to intercept (have special versions of) few collectd functions:
* metric submit: what metrics were reported or not reported, and with what exact values?
* logging: were warnings/errors generated as expected, or unexpectedly?
* config: were config option parsing results expected?

Unless you have a good suggestion on how to get access to submitted metrics and logs when using real collectd functions (without needing to implement both write & log plugins functionality), I think using build-internal archives would mess the tester.

> In other words, maintainability is more important to me than supporting such one-off use cases.

Yes that's understandable.

Skipping this PR is one option (for avoiding direct inclusion potentially causing later test breakages with unrelated collectd changes).

I would then need to update the mocked metric functions (removed in this PR) instead though.

Reply:
I thought about this a bit, because I believe that we need to improve our test coverage, particularly for frequently used plugins. My rough idea so far was:

* (Optional) Rename `plugin_mock.c` to `plugin_fake.c`
* In `src/daemon/plugin_mock.c` write a test-only implementation of `plugin_dispatch_metric_family()` that clones all metric families and stores the copy somewhere.
* Add functions for retrieving and resetting the stored metric families. Make these functions accessible to tests via `src/daemon/plugin_mock.h`.

This should allow (read) plugins to actually call the dispatch function successfully. Once the read function returns the test can inspect what metric families were dispatched, if any.

Possible alternatives:

* Implement a fake `plugin_register_write()` so that the fake `plugin_dispatch_metric_family()` call write callbacks without all the thread start-up required by the real `plugin.c`. Then provide a "write plugin for testing" implementation with the "copy and store in memory" behavior outlines above.
* Provide a test harness to do the above, but with the "real" `plugin.c` implementation. This would probably give the best test coverage, but is also the most complex to pull off.

We already have a fake logging implementation, we'd just need to also store log messages to make them available to the tests. This should be rather easy.

If the plugin is written so that it doesn't so any I/O in the config callback, the config parsing should be easy to test. 
#4238 is doing it, although with a temporary file for now. If we add the ability to parse a string to the config file code, then we should be able to test this without I/O also.

Reply:
**Config**

Sysman plugin test code calls registered config function for each key.   If real plugin code is used, it could call plugin config function directly instead.

**Logging**

Logs should not be much of a problem either.  The main thing (besides logs being printed for debugging) is able to get counts of things logged above given level, and being able to reset those counts.  With that, test can verify that failing things fail (with log output), and ones that should not fail, don't do that.

**Metrics**

This is the issue. Propsed thing will make debugging metric issues much harder.   My own fake code uses asserts on all possible errors, instead of just logging something, or exiting, like collectd code does.  Sysman plugin tester does a *lot* of things and backtraces from asserts right in the submit path, give a huge help for tracking down issues.  Backtraces do not help if metrics are collected first, and only then verified.

=> Better would be to have a test callback for metric submit

Reply:
I could avoid those includes by linking to `libcommon.la` and adding couple of dummy functions (for things that requires, but does not link).

Those includes are still in the code, but behind `BUILD_STANDALONE` ifdef, so that my own testing can do whole program analysis.